### PR TITLE
Suggest to add `move` keyword for generator capture

### LINKED ISSUE #64382: Consider async blocks on lifetime errors
[The following incorrect code](https://github.com/bluejekyll/trust-dns/pull/849/files#diff-5cbe22b68a836c7213c624099dedb640R182)
```rust
        // For QoS, this will only accept one message and output that
        // receive all inbound messages

        // TODO: this should match edns settings
        loop {
            let msg = if let Some(receiving) = receiving {
                // TODO: should we drop this packet if it's not from the same src as dest?
                let msg = ready!(receiving.as_mut().poll(cx))?;

                Some(Poll::Ready(Some(Ok(msg))))
            } else {
                None
            };
 
            *receiving = None;

            if let Some(msg) = msg {
                return msg;
            }

            let socket = Arc::clone(socket);
            let mut buf = [0u8; 2048];
            let receive_future = async {
                let socket = socket;

                let mut socket = socket.lock().await;
                let (len, src) = socket.recv_from(&mut buf).await?;
                
                Ok(SerialMessage::new(
                    buf.iter().take(len).cloned().collect(),
                    src,
                ))
            };

            *receiving = Some(Box::pin(receive_future));
        }
```
emits the following error
```
error[E0597]: `buf` does not live long enough
   --> crates/proto/src/udp/udp_stream.rs:204:56
    |
183 |               let msg = if let Some(receiving) = receiving {
    |                                                  --------- borrow later used here
...
200 |               let receive_future = async {
    |  ________________________________________-
201 | |                 let socket = socket;
202 | |
203 | |                 let mut socket = socket.lock().await;
204 | |                 let (len, src) = socket.recv_from(&mut buf).await?;
    | |                                                        ^^^ borrowed value does not live long enough
...   |
209 | |                 ))
210 | |             };
    | |_____________- value captured here by generator
...
213 |           }
    |           - `buf` dropped here while still borrowed
```

The correct code needs the `buf` to be part of the `async` block because it is a stack allocation that needs to be in the generator:

```rust
        loop {
            let msg = if let Some(receiving) = receiving {
                // TODO: should we drop this packet if it's not from the same src as dest?
                let msg = ready!(receiving.as_mut().poll(cx))?;

                Some(Poll::Ready(Some(Ok(msg))))
            } else {
                None
            };
 
            *receiving = None;

            if let Some(msg) = msg {
                return msg;
            }

            let socket = Arc::clone(socket);
            let receive_future = async {
                let socket = socket;
                let mut buf = [0u8; 2048];

                let mut socket = socket.lock().await;
                let (len, src) = socket.recv_from(&mut buf).await?;
                
                Ok(SerialMessage::new(
                    buf.iter().take(len).cloned().collect(),
                    src,
                ))
            };

            *receiving = Some(Box::pin(receive_future));
        }
```

Ideally, the error message would mention `buf`:

```
error[E0597]: `buf` does not live long enough
   --> crates/proto/src/udp/udp_stream.rs:204:56
    |
183 |               let msg = if let Some(receiving) = receiving {
    |                                                  --------- borrow later used here
...
199 |               let mut buf = [0u8; 2048];
    |                       --- consider moving this stack allocation inside the `async` block
200 |               let receive_future = async {
    |  ________________________________________-
201 | |                 let socket = socket;
202 | |
203 | |                 let mut socket = socket.lock().await;
204 | |                 let (len, src) = socket.recv_from(&mut buf).await?;
    | |                                                        ^^^ borrowed value does not live long enough
...   |
209 | |                 ))
210 | |             };
    | |_____________- value captured here by `async` block
...
213 |           }
    |           - `buf` dropped here while still borrowed
```

*thanks to @bluejekyll for bringing this up!*

CC @nikomatsakis

### PR DESCRIPTION
**Pull request by csmoe**
Closes #64382
r? @estebank

### CODE REVIEW

File: src/test/ui/async-await/async-borrowck-escaping-block-error.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// edition:2018
```
Comment (estebank) [inline src/test/ui/async-await/async-borrowck-escaping-block-error.rs]: Add `// run-rustfix` to make sure we never regress in the suggestion.

**Review by estebank — APPROVED** [review]

**Comment by estebank** [comment]
r=me when you're ready

**Comment by estebank** [comment]
I'll go ahead and mark this PR as ready for review so it doesn't come up as a draft.

@bors r+

File: src/test/ui/async-await/async-borrowck-escaping-block-error.stderr [hunk 2]
```diff
@@ -0,0 +1,22 @@
+error[E0373]: closure may outlive the current function, but it borrows `x`, which is owned by the current function
```
Comment (cramertj) [inline src/test/ui/async-await/async-borrowck-escaping-block-error.stderr]: Looks like the error is still referencing a "closure" here, which is unfortunate.

Comment (csmoe) [reply src/test/ui/async-await/async-borrowck-escaping-block-error.stderr]: @cramertj oops, my bad. will repair this soon.
but I'm not sure, as the [rfc](https://github.com/rust-lang/rfcs/blob/master/text/2394-async_await.md#async-blocks) said async block is equivalent to  immediately-invoked async closure.
would you mind suggesting a precise word here? cc @estebank

Comment (cramertj) [reply src/test/ui/async-await/async-borrowck-escaping-block-error.stderr]: "`async block` may outlive the current function"

**Pull request merged**